Parallel builds

OndroMih opened the pull request:
## Required

Associated issue: #2325

## Copyright acknowledgement

I acknowledge that the code will be owned by Manorrock.com and also the code in my future pull requests will be owned by Manorrock.com.

## Description

Several improvements to allow parallel builds with `mvn clean install -T 10` in the root project.

Most of the issues in the parallel builds are caused by:

* Builds of individual modules are slower because multiple builds are happening at the same time, eating a lot more CPU. Therefore standard timeouts are often not enough. Especially, Piranha Micro runtime takes longer to start. It's better to have some other waiting mechanism that doesn't rely on time. I implemented a mechanism that waits until Piranha port is occupied but some short wait time (1 second) is still necessary until Piranha is ready to serve requests.
* When running build in parallel, tests in the `test` directory are sometimes started in parallel to modules that build its dependencies. As a result, while a test is running, another build might be in the middle of building/deloying its artifact, and the tests tries to use a JAR/WAR in an inconsistent state. This often results in classes or resources not found on the classpath. Maven does a good job in resolving dependencies among parallel builds but sometimes it's necessary to add a dependency on the build module so that it's run before the test module that needs it.

I also randomized ports added a mechanism to find an available port so that multiple Piranha runtimes executed from parallel tests won't clash. This often wasn't an issue but it's a good practice anyway.

From now on, if we keep the following rules, builds will hopefully stay thread-safe:
* Always use the `PiranhaStartup` class in the `tests/common` module to wait for a Piranha runtime until it starts, don't wait for an explicit time period
* Always set the port with the `FreePortFinder` class in the `tests/common` module to avoid port clashes. The Arquillian connector already automatically searches for a free port by default so no need to do this for Arquillian-based tests.
* If tests fail on a class or resource not found on the classpath, investigate which Piranha runtime is used in the test and add it as a test dependency. Then run the failing module isolated and see if it passes the tests. 
* If the added dependency causes the test to fail (it's not expected to be on the classpath), remove it from the test classpath with the `classpathDependencyExcludes` configuration option in the `surefire` or `failsafe` plugin

OndroMih committed:
Improve stability of parallel builds
Wait for Piranha until it's ready - don't wait if fast, wait longer if slower (due to parallel execution)
Start the test module only after piranha artifacts are built - otherwise tests are run in parallel to building artifacts, leading to conflicts when reading and writing artifacts in local maven repo
Signed-off-by:Ondrej Mihályi <ondrej.mihalyi@gmail.com>

OndroMih committed:
Wait until Piranha starts with a check rather than a timeout
Signed-off-by:Ondrej Mihályi <ondrej.mihalyi@gmail.com>

OndroMih committed:
Further improvements to parallel builds.
Automatic port assignment for Arquillian containers by default.
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

OndroMih committed:
Further improvements to parallel builds. Allow building offline.
Finer grained dependencies between test modules and build modules to avoid polluting test classpath.
Helloworld test module binds to a random port.
Waiting for Piranha startup now waits additional 1 second to give some time for Piranha to become ready after it binds to the port.
Piranha Maven plugin catches all IO exceptions when trying to download a ZIP file from Maven Central. This allows running the build also without a network connection.
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

OndroMih commented:
I executed `mvn clean install -T 10` in the root directory 5 times in a row and the build succeeded in all 5 cases. Before, I got a failure in almost every build. So I believe that building in parallel is now pretty stable.

My CPU has 8 cores, most of the time all of them were 100% utilized. With this PR, build time with `mvn clean install -T 10` on my laptop is 3 minutes 51 seconds. Without `-T 10`, it takes 8 minutes 18 seconds to build.

OndroMih committed:
Clean up
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

Thihup commented on test/common/src/main/java/module-info.java:
```diff
@@ -0,0 +1,33 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+module cloud.piranha.arquillian.server {
```
Probably this module name should be `cloud.piranha.test.common`

OndroMih replied:
Yes, a copy & paste error.

Thihup commented on test/common/src/main/java/cloud/piranha/test/common/FreePortFinder.java:
```diff
@@ -0,0 +1,77 @@
+ *
+ * Finds a free port that isn't currently bound to by anything else.
+ */
+public class FreePortFinder {
```
In a few places, we're using the free-port-finder test dependency. Would it be possible to remove that dependency using this class?

OndroMih replied:
I didn't notice that. yes, I can use that dependency. My solution selects candidate ports randomly to avoid clashes if 2 threads are searching for a port at the same time and both would find the same free port while I don't see that the free-port-finder would do that (https://github.com/alexpanov/freeportfinder/blob/master/src/main/java/me/alexpanov/net/FreePortFinder.java) but it looks like it has some logic to prevent clashes. Let's see if it's enough.

OndroMih replied:
But I'll leave my free port finder logic in the Arquillian container, I don't want to bring another dependency to the Arquillian container just for that. Is that OK?

Thihup replied:
Sounds good! 😉

Thihup commented on micro/shrinkwrap/loader/src/main/java/cloud/piranha/micro/shrinkwrap/loader/MicroConfiguration.java:
```diff
@@ -234,7 +234,7 @@ public Map<String, Object> toMap() {
     }
 
     /**
-     * {@return the version}
```
It seems that removing the braces changes a little the Javadoc.

![image](https://user-images.githubusercontent.com/13357965/170881441-1f68c00f-3bfb-4fd1-8971-c62332595ac4.png)

OndroMih replied:
I don't mind reverting this, it looked to me that the version with braces was ignored, but it was probably ignored only by Netbeans when I previewed the javadoc. I didn't actually know that the vesion with braces is a valid javadoc.

Thihup reviewed (changes requested):
Nice work!
Overall, LGTM, just a few comments 😉

OndroMih committed:
Edits after code review https://github.com/piranhacloud/piranha/pull/2584
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

mnriem commented:
Hi @OndroMih 

As this is your first PR can you please acknowledge that the code will be owned by Manorrock.com. I will only ask this one time as that is all that is needed for contributing code back to us.

mnriem commented:
@OndroMih some minor things for future PRs. We try not to introduce whitespace changes as much as possible as it generates a lot of noise when doing PRs. And try to add JavaDoc to every public method / class to describe what it does. Other than that this is looking good. Just waiting for the acknowledgement and then we can merge this PR

mnriem reviewed (changes requested):
As this is your first PR can you please acknowledge that the code will be owned by Manorrock.com. I will only ask this one time as that is all that is needed for contributing code back to us.

OndroMih commented:
Hi, @mnriem , I've added a section "Copyright acknowledgement" into the description of this PR, is it enough?

OndroMih committed:
Updates from code review 
Issue: https://github.com/piranhacloud/piranha/issues/2325
PR: https://github.com/piranhacloud/piranha/pull/2584
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

OndroMih committed:
Arquillian connector checks whether it can bound to the found free port.
Some more documentation.
Issue: https://github.com/piranhacloud/piranha/issues/2325
PR: https://github.com/piranhacloud/piranha/pull/2584
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

OndroMih commented:
I've added some more documentation and improved the free port finder to check whether it can bind to the free port before it returns it.

OndroMih committed:
Edited comments to please Checkstyle.
Issue: https://github.com/piranhacloud/piranha/issues/2325
PR: https://github.com/piranhacloud/piranha/pull/2584
Signed-off-by:Ondrej Mihalyi <ondrej.mihalyi@gmail.com>

mnriem commented:
@OndroMih Yea that is sufficient. Looking forward to more of your contributions. And welcome!

mnriem commented:
@OndroMih Let us know when the PR is ready :)

OndroMih commented:
The PR is ready now, you can merge it if you want. Even my last tweaks weren't necessary but now there's really nothing to add.

And thanks for your welcome :) Piranha is an interesting project and I'm glad I can contribute to it.

mnriem commented:
@OndroMih Congratz on your first PR. Would you mind if we announce it on Twitter?

OndroMih commented:
Yes, feel free to announce it. I have something to celebrate :)
